### PR TITLE
Optimal rearrangment

### PR DESCRIPTION
**Pull request by fulopbbea**
-added method for a greedy algorithm that (tries to) find the optimal rearrangement (least no. of moves needed)
    -it moves the car that should go in the current place of the empty slot, when it's possible
    -when it's not possible (the empty slot is in the right place), it moves the first misplaced car to the empty slot to deblock the procedure
-slightly enhanced the move() method (less error prone)
-added some test cases for the new algorithm

!!!!!!!!!!!!!!!!!!!!!!!!!!!! TO BE DONE !!!!!!!!!!!!!!!!!!!!!!!!!!!!!
-add test cases that cover every bit of the code
-examine if the solutions are actually optimal or not
-!!!read through and enhance the comments please
-MAYBE add methods to automatically test the optimality of the solutions

### CODE REVIEW

**Comment by fulopbbea** [comment]
forgot to mention, I've implemented it the simplest way possible, it's still possible to rewrite it using the above mention queue/list to make the complexity smaller

File: Assignment6.java [hunk 1]
```diff
@@ -31,36 +34,87 @@
             // If i isn't already empty.
             if (indexMap.get(0) != i) {
                 // Move the number in i to the empty slot so that i becomes empty.
-                Move emptyCurrentSlot = new Move(given[i], i, indexMap.get(0));
-                
-                listOfMoves.add(emptyCurrentSlot);
-                move(emptyCurrentSlot, given, indexMap);
+                listOfMoves.add(move(given[i], given, indexMap));
```
Comment (cba33) [inline Assignment6.java:38]: I have the impression we don't need indexMap. Instead, we could simply compare given[i] with desired[i].

File: Assignment6.java [hunk 2]
```diff
@@ -31,36 +34,87 @@
...
+        }
+        
+        //car belonging to the current empty slot
+        if (desired[indexMap.get(0)] != 0) {
+            return desired[indexMap.get(0)];
+        }
+        
+        //empty lot in its desired place => deblock the algorithm
+        //find the first car that isn't in its right place
+        for (int i = 0; i < given.length; i++) {
```
Comment (cba33) [inline Assignment6.java]: If we unblock cars in the order given by the array 'desired', then we know we never need to go back to those positions, so we wouldn't need to loop all cars again. This becomes interesting when we have a huge array with just a few misplaced cars.

**Review by cba33 — CHANGES_REQUESTED** [review]
Looks good to me! A few comments:

File: Assignment6.java (hunk 1 above)
Comment (cba33) [reply Assignment6.java:38]: On a second thought, the map is actually useful :D Ignore this one.

File: Assignment6.java (hunk 2 above)
Comment (fulopbbea) [reply Assignment6.java]: If I understand correctly what you say, wouldn't we need some kind of global variable to know where to start the search next time?

Comment (cba33) [reply Assignment6.java]: Indeed, we'll use the queue / set as a global variable holding misplaced cars.

File: Assignment6.java [hunk 3]
```diff
@@ -77,24 +77,28 @@
...
+
         //empty lot in its desired place => deblock the algorithm
-        //find the first car that isn't in its right place
-        for (int i = 0; i < given.length; i++) {
-            if (given[i] != 0 && given[i] != desired[i]) {
-                return given[i];
-            }
+        // pick a random car from set of misplaced cars
+        if (misplacedCars.size() > 0) {
+            Integer nextCarToMove = (Integer)misplacedCars.toArray()[0];
```
Comment (fulopbbea) [inline Assignment6.java]: Isn't toArray() O(n)?

File: Assignment6.java [hunk 4]
```diff
@@ -77,24 +77,28 @@
...
         }
-        
+
         //empty lot in its desired place => deblock the algorithm
-        //find the first car that isn't in its right place
-        for (int i = 0; i < given.length; i++) {
-            if (given[i] != 0 && given[i] != desired[i]) {
-                return given[i];
-            }
+        // pick a random car from set of misplaced cars
```
Comment (elifguler96) [inline Assignment6.java]: why did you decide to pick the car randomly instead of picking it in the order of the desired array?

File: Assignment6.java [hunk 5]
```diff
@@ -77,24 +77,28 @@
...
+        if (misplacedCars.size() > 0) {
+            Integer nextCarToMove = (Integer)misplacedCars.toArray()[0];
+            misplacedCars.remove(nextCarToMove);
```
Comment (fulopbbea) [inline Assignment6.java]: I guess we shouldn't remove the car yet as it still isn't in its right place. We move it where the empty slot should be.

File: Assignment6.java [hunk 6]
```diff
@@ -1,9 +1,6 @@
 package assignment6;
 
-import java.util.HashMap;
-import java.util.LinkedList;
-import java.util.List;
-import java.util.Map;
+import java.util.*;
```
Comment (fulopbbea) [inline Assignment6.java]: We should only add the libs we actually use.

**Review by fulopbbea — COMMENTED** [review]
Good idea with the HashSet :) and nice clean code in general. :D I'm only really concerned about that toArray()'s complexity which would mean we haven't reduced the complexity at all.

File: Assignment6.java (hunk 4 above)
Comment (cba33) [reply Assignment6.java]: Because I think it leads to the same number of moves (I think someone else commented that on slack too, and I agree). At least on the test examples I quickly ran it on. Did you find an example where we observe a better behavior?

File: Assignment6.java (hunk 3 above)
Comment (cba33) [reply Assignment6.java]: Good point, though the amortized complexity should be lower when we have a huge number of cars and just a few misplaced ones. If we change to a list / queue, removing elements from the list every time we put them in place will be O(N), as opposed to removing from a set. Any ideas?

File: Assignment6.java (hunk 5 above)
Comment (cba33) [reply Assignment6.java]: Good point, I agree. I'll quickly make the change.

File: Assignment6.java (hunk 6 above)
Comment (cba33) [reply Assignment6.java]: I'll make the change.

**Comment by fulopbbea** [comment]
How about we write and executeMoves() helper method in the test class and add some tests where we test if the generated list of moves actually transforms the given array into the desired one? :) As we update the algorithm it gets kinda gruesome to check and possibly update the test cases. :/

File: Assignment6.java (hunk 3 above)
Comment (fulopbbea) [reply Assignment6.java]: It's clearly better then the previous version and using any kind of data structure I can think of we increase either the delete or the search time so I guess it works for me. :)

File: Assignment6.java [hunk 7]
```diff
@@ -31,36 +35,93 @@
...
+        if (desired[indexMap.get(0)] != 0) {
+            return desired[indexMap.get(0)];
+
+        }
+
+        //empty lot in its desired place => deblock the algorithm
+        // pick a random car from set of misplaced cars
+        if (misplacedCars.size() > 0) {
+            Integer nextCarToMove = misplacedCars.iterator().next();
+            misplacedCars.remove(nextCarToMove);
```
Comment (fulopbbea) [inline Assignment6.java:101]: I guess we still shouldn't remove this car yet as it is not in its right place. Or am I missing something? :/

**Review by fulopbbea — COMMENTED** [review]
Really creative idea. :D

File: Assignment6.java (hunk 7 above)
Comment (elifguler96) [reply Assignment6.java:101]: I think so too

Comment (katermar) [reply Assignment6.java:101]: I think the same, similarly as in the previous version

File: Assignment6.java [hunk 8]
```diff
@@ -94,8 +172,7 @@ public static void printListOfMoves(List<Move> plan) {
      * @param m the move
      */
     private static void printMove(Move m) {
-        System.out.printf("car %d moves from %d to %d", m.getCar(), m.getFrom(), m.getTo());
```
Comment (katermar) [inline Assignment6.java:97]: Don't we still need to print all these messages?

File: Assignment6Test.java [hunk 9]
```diff
@@ -70,4 +73,73 @@ public void rearrangeTestShift() {
...
+    public void rearrangeGreedyTestArranged() {
+        int[] given = {1, 2, 0, 3};
+        int[] desired = {1, 2, 0, 3};
+
+        List<Move> list = rearrangeGreedy(given, desired);
+        assertEquals(list, Collections.emptyList());
+    }
+
+    @Test
+    public void rearrangeGreedyTestEmpty() {
```
Comment (katermar) [inline Assignment6Test.java:114]: There also can be a test for given or desired empty array. (Maybe it makes no sence)

File: Assignment6Test.java [hunk 10]
```diff
@@ -70,4 +73,73 @@ public void rearrangeTestShift() {
...
+    @Test
+    public void rearrangeGreedyLongArray() {
+        int[] given = {1, 2, 3, 4, 5, 6, 0};
+        int[] desired = {3, 1, 6, 2, 4, 5, 0};
+
+        List<Move> list = rearrangeGreedy(given, desired);
+        // check we perform 7 moves
+        assertEquals(list.size(), 7);
+    }
+
```
Comment (katermar) [inline Assignment6Test.java:144]: Maybe we can have a test for a very small array like 1 or 2 cars?

**Review by katermar — APPROVED** [review]
Looks really awesome! Good job 👍 
Superfine idea with an iterator 🥇

File: Assignment6.java (hunk 8 above)
Comment (elifguler96) [reply Assignment6.java:97]: We do print the message, it is two lines below 😄

**Review by elifguler96 — APPROVED** [review]
